### PR TITLE
#1379: added ntdll method for suspending/resuming processes

### PR DESCRIPTION
**Pull request by ogurets**
Slightly modified patch for #1379, which will use threading method if `NtSuspendProcess` or `NtResumeProcess` are unavailable.

### CODE REVIEW

File: psutil/_psutil_windows.c [hunk 1]
```diff
@@ -1208,6 +1209,48 @@ psutil_proc_suspend_or_resume(DWORD pid, int suspend) {
...
+    }
+
+    hProcess = OpenProcess(PROCESS_SUSPEND_RESUME, FALSE, pid);
+    if (hProcess == INVALID_HANDLE_VALUE) {
+        PyErr_SetFromWindowsErr(0);
+        return FALSE;
+    }
+
+    if (suspend == 1) {
+        if (pfnNtSuspendProcess(hProcess) == (DWORD) - 1) {
```
Comment (giampaolo) [inline psutil/_psutil_windows.c]: I think it's safe (and cleaner) to just do:

```c
if (pfnNtSuspendProcess(hProcess) == - 1) {
```

File: psutil/_psutil_windows.c [hunk 2]
```diff
@@ -1208,6 +1209,48 @@ psutil_proc_suspend_or_resume(DWORD pid, int suspend) {
...
+            PyErr_SetFromWindowsErr(0);
+            CloseHandle(hProcess);
+            return FALSE;
+        }
+    } else {
+        if (pfnNtResumeProcess(hProcess) == (DWORD) - 1) {
```
Comment (giampaolo) [inline psutil/_psutil_windows.c]: same here (remove DWORD casting)

**Review by giampaolo — CHANGES_REQUESTED** [review]
Nice! I added some comments inline. Other than that could you please limit lines length to 80 chars?

**Comment by EccoTheFlintstone** [comment]
According to this : 
https://social.msdn.microsoft.com/Forums/vstudio/en-US/e119f8e8-98bc-466e-a9bb-88bcbde6524c/why-does-ntsuspendprocess-fail-with-processsuspendresume-but-succeeds-with-processallaccess?forum=windowssdk

`NtSuspendProcess` and `NtResumeProcess` return a `NTSTATUS`, so 0 == success, anything else == failure, no ?

Moreover, `psutil_handle_from_pid `seems to always need `PROCESS_QUERY_INFORMATION / PROCESS_QUERY_LIMITED_INFORMATION`, as it calls `psutil_check_phandle` which calls `psutil_is_phandle_running` which calls `GetExitCodeProcess` which needs `PROCESS_QUERY_INFORMATION `(https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-getexitcodeprocess)

The problem has not arisen before as, currently, every call to  `psutil_handle_from_pid` want `PROCESS_QUERY_INFORMATION or PROCESS_QUERY_LIMITED_INFORMATION`

**Comment by ogurets** [comment]
Should I add these flags into `psutil_handle_from_pid_waccess` then?
And what about `PROCESS_VM_READ`? Why is it switched on at `psutil_handle_from_pid` if we don't need it?

File: psutil/_psutil_windows.c [hunk 3]
```diff
@@ -1208,6 +1209,48 @@ psutil_proc_suspend_or_resume(DWORD pid, int suspend) {
...
+{
+    HANDLE hProcess = NULL;
+
+    typedef LONG (NTAPI *NtSuspendProcess)(IN HANDLE ProcessHandle);
+    typedef LONG (NTAPI *NtResumeProcess)(IN HANDLE ProcessHandle);
+    NtSuspendProcess pfnNtSuspendProcess = (NtSuspendProcess)GetProcAddress(GetModuleHandle("ntdll"), "NtSuspendProcess");
+    NtResumeProcess pfnNtResumeProcess = (NtResumeProcess)GetProcAddress(GetModuleHandle("ntdll"), "NtResumeProcess");
+    if (!pfnNtSuspendProcess || !pfnNtResumeProcess) {
+        // Use alternate method
+        return psutil_proc_suspend_or_resume_threads(pid, suspend);
```
[the remark this answers is not in the document]
Comment (ogurets) [reply psutil/_psutil_windows.c]: Done.

**Comment by EccoTheFlintstone** [comment]
Slight typo here : 
you should reverse condition check (STATUS_SUCCESS == 0, error if != 0)

```
if (pfnNtSuspendProcess(hProcess) != STATUS_SUCCESS) {...
..
if (pfnNtResumeProcess(hProcess) != STATUS_SUCCESS) { ...

```
As for the `psutil_handle_from_pid` part, it's intended more for @giampaolo than you :)
But it seems that your fork is outdated, because `psutil_handle_from_pid_waccess` doesn't exist anymore and `psutil_handle_from_pid` doesn't use default rights (`QUERY_INFORMATION | VM_READ`) anymore either (patch was commited a few days ago)

**Comment by ogurets** [comment]
@EccoTheFlintstone, that's unfortunate, I branched from `5.4.8` tag intentionally to fix bugs while not pulling untested features from `master`.
Would it be better to pull the latest `master` and merge it into my fork's branch to keep the pull request or should I close it and make a new one?

**Comment by EccoTheFlintstone** [comment]
I'm no git expert, but doing a new pull request might be the cleanest method (least changes to merge)

As for the `psutil_handle_from_pid` part (with the rights), might be better to wait for @giampaolo  advice  as to how to proceed

**Comment by ogurets** [comment]
@EccoTheFlintstone, OK, I've pushed my git knowledge to the limit and after some crazy rebasing, picking and amending have managed to clear the mess (I hope).
Created another request to save time in case @giampaolo chooses to go on from there.

**Comment by giampaolo** [comment]
Sorry for being slow at responding on this. If you're still interested I'd like to push this forward. Some considerations about your patch:

- it appears you added new C code which is not used 
- I don't think it's worth it to keep suspend/resume support for Windows XP, therefore you can remove the old implementation relying on `CreateToolhelp32Snapshot` 
- on Windows XP it's fine for `suspend()` and `resume()` to raise `NotImplementedError`; you can do it in C as `PyErr_SetString(PyExc_NotImplementedError, "Windows versions too old");`
- you can move the GetProcAddress / GetModuleHandle logic in here:
https://github.com/giampaolo/psutil/blob/master/psutil/arch/windows/global.c#L131
I just added this stuff which basically loads these libs on startup (and never again). You'll have to define the type in `ntext.h` first (not tested):
```c
typedef NTSTATUS (WINAPI *_pfnNtSuspendProcess) (
    HANDLE hProcess
);
```
Please let me know if you're willing to work on this, else I will do it (I intend to publish a new version soon now that I have some spare time).

Extra: I changed a lot of stuff on Windows and am not sure if I broke Windows XP support so your testimony about current GIT master status would be very useful.

**Comment by ogurets** [comment]
Actually, the next PR was more "fresh" - https://github.com/giampaolo/psutil/pull/1383 as it was rebased onto some latest changes you had made to date.

Sadly, now I'm really busy with my job and wouldn't be able to contribute seriously.
Plus, my dev interest is WinXP + stable versions of Debian Linux and I'm not planning to use later versions of Windows ever (slowly migrating to Linux completely), therefore I won't be of much use developing or testing Vista+ stuff.

The good news: as I'm using `psutil` daily in my personal "domestic" application, I'll have to maintain support of WinXP in my fork (at least) and will probably offer you more PRs in future.

**Comment by giampaolo** [comment]
OK, I will go on and do it myself then, thanks. When you have time, could you please verify that current GIT master still works on Win XP?

**Comment by giampaolo** [comment]
Implemented in #1435. Thanks for the idea. Closing this one as a duplicate.

**Pull request closed without merging**